azure/services/scalesets/spec.go

```diff
@@ -0,0 +1,522 @@
+	}
+
+	if s.AcceleratedNetworking == nil {
+		// set accelerated networking to the capability of the VMSize
+		accelNet := s.SKU.HasCapability(resourceskus.AcceleratedNetworking)
+		s.AcceleratedNetworking = &accelNet
+	}
+
+	extensions, err := s.generateExtensions(ctx)
+	if err != nil {
+		return compute.VirtualMachineScaleSet{}, err
+	}
+
+	storageProfile, err := s.generateStorageProfile(ctx)
+	if err != nil {
+		return compute.VirtualMachineScaleSet{}, err
+	}
+
+	securityProfile, err := s.getSecurityProfile()
+	if err != nil {
+		return compute.VirtualMachineScaleSet{}, err
+	}
+
+	priority, evictionPolicy, billingProfile, err := converters.GetSpotVMOptions(s.SpotVMOptions, s.OSDisk.DiffDiskSettings)
+	if err != nil {
+		return compute.VirtualMachineScaleSet{}, errors.Wrapf(err, "failed to get Spot VM options")
+	}
+
+	diagnosticsProfile := converters.GetDiagnosticsProfile(s.DiagnosticsProfile)
+
+	osProfile, err := s.generateOSProfile(ctx)
+	if err != nil {
+		return compute.VirtualMachineScaleSet{}, err
+	}
+
+	orchestrationMode := converters.GetOrchestrationMode(s.OrchestrationMode)
+
+	vmss := compute.VirtualMachineScaleSet{
+		Location: pointer.String(s.Location),
+		Sku: &compute.Sku{
+			Name:     pointer.String(s.Size),
+			Tier:     pointer.String("Standard"),
+			Capacity: pointer.Int64(s.Capacity),
+		},
+		Zones: &s.FailureDomains,
+		Plan:  s.generateImagePlan(ctx),
+		VirtualMachineScaleSetProperties: &compute.VirtualMachineScaleSetProperties{
+			OrchestrationMode:    orchestrationMode,
+			SinglePlacementGroup: pointer.Bool(false),
+			VirtualMachineProfile: &compute.VirtualMachineScaleSetVMProfile{
+				OsProfile:          osProfile,
+				StorageProfile:     storageProfile,
+				SecurityProfile:    securityProfile,
+				DiagnosticsProfile: diagnosticsProfile,
+				NetworkProfile: &compute.VirtualMachineScaleSetNetworkProfile{
+					NetworkInterfaceConfigurations: s.getVirtualMachineScaleSetNetworkConfiguration(),
+				},
+				Priority:       priority,
+				EvictionPolicy: evictionPolicy,
+				BillingProfile: billingProfile,
+				ExtensionProfile: &compute.VirtualMachineScaleSetExtensionProfile{
+					Extensions: &extensions,
+				},
+			},
+		},
+	}
+
+	// Set properties specific to VMSS orchestration mode
+	switch orchestrationMode {
+	case compute.OrchestrationModeUniform:
+		vmss.VirtualMachineScaleSetProperties.Overprovision = pointer.Bool(false)
+		vmss.VirtualMachineScaleSetProperties.UpgradePolicy = &compute.UpgradePolicy{Mode: compute.UpgradeModeManual}
+	case compute.OrchestrationModeFlexible:
+		vmss.VirtualMachineScaleSetProperties.VirtualMachineProfile.NetworkProfile.NetworkAPIVersion =
+			compute.NetworkAPIVersionTwoZeroTwoZeroHyphenMinusOneOneHyphenMinusZeroOne
+		vmss.VirtualMachineScaleSetProperties.PlatformFaultDomainCount = pointer.Int32(1)
+		if len(s.FailureDomains) > 1 {
+			vmss.VirtualMachineScaleSetProperties.PlatformFaultDomainCount = pointer.Int32(int32(len(s.FailureDomains)))
+		}
```

Comment:
Can we point to the source of these rules in a comment?

Reply:
I'm not sure I understand. Do you want to link to this block in another section of the PR?

Reply:
Sorry for the confusion. I was suggesting if we could point the docs for these validations via comment. Something like `For rules, refer: azure-docs.com`. <- Something along the lines so that its easier to review the rules next time we come across this switch case

Reply:
@nawazkh Ah gotcha. I'm not sure where to find these rules from. @CecileRobertMichon any thoughts?

Reply:
AFAIK these have to do with differences between flex and uniform VMSS. @mboersma might know more.

Reply:
I honestly don't know where these values originate. I inherited them from @devigned's original work to support VMSS Flex and learned through experimentation that they were indeed necessary.
If we do want to try to link to a doc, I think https://learn.microsoft.com/azure/virtual-machine-scale-sets/virtual-machine-scale-sets-orchestration-modes is good and may explain some of these parameters.

Reply:
Cool I'll drop the link in a comment.